### DISCUSSION TITLE
Add Windows on ARM build

### DISCUSSION THREAD

**Discussion by hmartinez82** [mamedev/discussions#112] [Ideas]
It should be straightforward to build for ARM64. Unless some third party library does not support it yet.

MSYS2 has the CLANGARM64 environment.

**Comment by rb6502** [comment]
We've been waiting for official GCC MinGW-ARM support, which has been announced recently.  ARM and AArch64 builds on macOS and Linux are already supported of course, so it should be straightforward.

**Comment by talynone** [comment]
@rb6502 Will this help?

GitHub Actions now supports Windows on Arm runners for all public repos:

https://blogs.windows.com/windowsdeveloper/2025/04/14/github-actions-now-supports-windows-on-arm-runners-for-all-public-repos/

**Reply by hmartinez82** [reply]
Yes! It should. I'm going to try building for ARM64.

@rb6502 mame already supports being built with CLANG. I looked at the `ci-windows.yml` workflow and I noticed that mame is being built with gcc and clang, but it's built with clang in MSYS2's MINGW64 environment. That's an interesting thing I haven't seen yet:) . Normally we build with GCC in the MINGW64 env and with Clang in the CLANG64 env.

I'm going to try building mame in the CLANGARM64 in my ARM64 laptop, if that works then I'll create a pr for using the new runner that @talynone posted about above.

**Reply by hmartinez82** [reply]
Success!
![image](https://github.com/user-attachments/assets/18bfae9c-e8d3-49b6-9622-2717ceee512d)

![image](https://github.com/user-attachments/assets/a97b378d-4f8e-4716-a09b-497a255b17a2)

**Reply by cuavas** [reply]
Try something that uses a recompiler.

**Reply by hmartinez82** [reply]
@cuavas You mean like a game? Which one can I use? Sorry, I'm not a ROM expert :(

**Reply by cuavas** [reply]
Something with a Hitachi SH-2, Hitachi SH-4, Hyperstone E1, IBM PowerPC or MIPS III CPU.

**Reply by hmartinez82** [reply]
@cuavas How can I do that? Can you point me to the ROM?

**Reply by hmartinez82** [reply]
@cuavas I found one rom for IBM PowerPC and one for Hitachi SH-2. They both SEGFAULT with:
```
-----------------------------------------------------
Exception at EIP=00007FF7C0324E14 (not found): ACCESS VIOLATION
While attempting to read memory at 0000000000000028
-----------------------------------------------------
 X0=000000FAA09CA030  X1=0000000000000001  X2=0000021395EE8470  X3=0000000000000002
 X4=00000000FFFFFFFF  X5=00007FF7C6862660  X6=00007FF7C6862670  X7=0000000000000000
 X8=0000000000000213  X9=0000000000000000 X10=0000000000000000 X11=0000000000000000
X12=0000000000000000 X13=0000000000000020 X14=0000000000000000 X15=000000000000000F
X16=00000000FFFFFFF0 X17=0000000000000000 X18=11277FFB07050128 X19=0000021395EE8470
X20=00000213BBD60000 X21=00007FF7C0324D2C X22=000000FAA09C9D90 X23=000000FAA09C9D28
X24=00000000000000FF X25=0DE0B6B3A763FFFF X26=000000003B9AC9FF X27=0000000000002BC0
X28=000000000000E90B  FP=00007FF7C579DF80  LR=00007FF7C030BD54  SP=000000FAA09C9AC0
-----------------------------------------------------
Stack crawl:
  00007FF7C579DF80: 00007FF7C0324E14 (not found)
  00007FF7C579DF80: 00007FF7C030BD54 (not found)
  00007FF7C579DF80: 00007FF7BD65689C (not found)
  00007FF7C579DF80: 00007FF7BCD3E7A4 (not found)
  00007FF7C579DF80: 00007FF7BCD3EC38 (not found)
  00007FF7C579DF80: 00007FF7BC08EBB0 (not found)
  00007FF7C579DF80: 00007FF7BC0A19C8 (not found)
  00007FF7C579DF80: 00007FF7BE6750B8 (not found)
  000000FAA09CFF60: 00007FF7C15A71F8 (not found)
  000000FAA09CFF60: 00007FF7C15A8320 (not found)
  000000FAA09CFF60: 00007FF7BE675EC0 (not found)
  000000FAA09CFF60: 00007FF7BBEB16E4 (not found)
  000000FAA09CFF60: 00007FF7B8B113C8 (not found)
  000000FAA09CFF60: 00007FF7B8B1142C (not found)
  000000FAA09CFF60: 00007FFBD0B18740 (BaseThreadInitThunk+0x0040)
  000000FAA09CFF70: 0606FFFBD1CAD834 (not found)
Segmentation fault
```

**Comment by hmartinez82** [comment]
Here's a first go https://github.com/mamedev/mame/pull/13597

**Comment by hmartinez82** [comment]
Here's the output now of a DEBUG build with symbols:
```
-----------------------------------------------------
Exception at EIP=00007FF62B7156C4 (_ZNK3drc12_GLOBAL__N_19drcbe_x6411mov_r64_immERN6asmjit9_abi_1_133x869AssemblerERKNS4_2GpEy+0x008c): ACCESS VIOLATION
While attempting to read memory at 0000000000000028
-----------------------------------------------------
 X0=000000D1632F9D30  X1=0000000000000001  X2=000001FAEFE3DE90  X3=0000000000000002
 X4=00000000FFFFFFFF  X5=00007FF632223CA0  X6=00007FF632223CB0  X7=0000000000000001
 X8=00000000000001FA  X9=0000000000000000 X10=0000000000000000 X11=0000000000000000
X12=0000000000000000 X13=0000000000000000 X14=0000000000000000 X15=0000000000000000
X16=00007FFBCDAF30C0 X17=000000D1632F9D84 X18=000001FAF8046000 X19=000001FAEFE3DE90
X20=000001FB14D80000 X21=00007FF62B7155D8 X22=000000D1632F9A90 X23=000000D1632F9A28
X24=00000000000000FF X25=000000D1632FF428 X26=0DE0B6B3A763FFFF X27=00007FF6310687D8
X28=F21F494C589C0000  FP=000000D1632F97E0  LR=00007FF62B6FC81C  SP=000000D1632F97C0
-----------------------------------------------------
Stack crawl:
  000000D1632F97E0: 00007FF62B7156C4 (_ZNK3drc12_GLOBAL__N_19drcbe_x6411mov_r64_immERN6asmjit9_abi_1_133x869AssemblerERKNS4_2GpEy+0x008c)
  000000D1632F9FD8: 00007FF62B6FC81C (_ZN3drc12_GLOBAL__N_19drcbe_x645resetEv+0x04a8)
  000000D1632F9FF8: 00007FF6286C9E88 (_ZN12drcuml_state5resetEv+0x004c)
  000000D1632FA018: 00007FF627FBC5F0 (_ZN19sh_common_execution16code_flush_cacheEv+0x0018)
  000000D1632FA038: 00007FF627FBCB20 (_ZN19sh_common_execution15execute_run_drcEv+0x0024)
  000000D1632FA070: 00007FF62B3B9210 (_ZN10sh2_device11execute_runEv+0x002c)
  000000D1632FA080: 00007FF62B3B9398 (_ZThn600_N10sh2_device11execute_runEv+0x0010)
  000000D1632FA110: 00007FF626E9EBBC (_ZN16device_scheduler9timesliceEv+0x0388)
  000000D1632FA228: 00007FF626EB2704 (_ZN15running_machine3runEb+0x03a4)
  000000D1632FF6A0: 00007FF629942B10 (_ZN20mame_machine_manager7executeEv+0x0160)
  000000D1632FF858: 00007FF62CB2936C (_ZN12cli_frontend15start_executionEP20mame_machine_managerRKNSt3__16vectorINS2_12basic_stringIcNS2_11char_traitsIcEENS2_9allocatorIcEEEENS7_IS9_EEEE+0x0180)
  000000D1632FF988: 00007FF62CB2A518 (_ZN12cli_frontend7executeERNSt3__16vectorINS0_12basic_stringIcNS0_11char_traitsIcEENS0_9allocatorIcEEEENS5_IS7_EEEE+0x004c)
  000000D1632FF9C8: 00007FF629943A98 (_ZN13emulator_info14start_frontendER11emu_optionsR13osd_interfaceRNSt3__16vectorINS4_12basic_stringIcNS4_11char_traitsIcEENS4_9allocatorIcEEEENS9_ISB_EEEE+0x0030)
  000000D1632FFCF0: 00007FF626C9D434 (main+0x02bc)
  000000D1632FFD60: 00007FF622A013C8 (__tmainCRTStartup+0x0210)
  000000D1632FFD60: 00007FF622A0142C (mainCRTStartup+0x0014)
  000000D1632FFD60: 00007FFBD0B18740 (BaseThreadInitThunk+0x0040)
  000000D1632FFD70: FC3AFFFBD1CAD834 (wcsncmp+0xa3dbb29c)
Segmentation fault
```

**Reply by hmartinez82** [reply]
@cuavas It seems that the code is being JIT'ed to x86_64 given that `drcbe_x64` showed up in the stack . I'm going to try forcing the `PLATFORM` value  in the Makefile to `arm64`. Crossing fingers that will do it.

**Reply by hmartinez82** [reply]
That did it!
![image](https://github.com/user-attachments/assets/609c0968-8a07-44ca-b987-07eec161a04e)
![image](https://github.com/user-attachments/assets/b4285396-d429-48fe-932a-445ddeac39e8)

**Comment by talynone** [comment]
The ARM build is now part of the automated build. But it's not making it through the releases section which seems to host some binary that goes through some secondary unknown re-packaging into some installer .exe with extra executables/folders. This installer is x64 only.

**Reply by cuavas** [reply]
Yeah, releases are built a bit differently to CI builds and subject to more testing.  If you want to buy me a nice ARM notebook, I can put more effort into ARM support in general (including the 64-bit ARM recompiler, etc.).

**Comment by rb6502** [comment]
This is done now - we now post official Windows-on-ARM binaries and recompiler support for AArch64 is much improved.

**Reply by cuavas** [reply]
To add to this, 64-bit ARM is basically a “tier one” supported platform now, including all the “nice to have” things, as far as I know:
* Native recompiler back-end
* eminline.h specialisations
* Full precision bilinear filtering implementation
* CI builds for macOS and Windows
* Windows binary release packages
* “Arm64” choice for build type at MAME Testers

Any remaining AArch64-specific issues probably just come down to people writing bad code.  The ones we’ve fixed so far (that I can remember) fall into these categories:
* Playing fast and loose with integer widening/narrowing
* Relying on semantics of floating point to integer conversion when the input is out of range
* Relying on total store order semantics
* Lack of proper thread synchronisation causing race conditions
